pkg/kubelet/images/image_pull_manager.go

```diff
@@ -0,0 +1,663 @@
+		// anyone on this node can access the image
+		return false
+	}
+
+	if len(cachedCreds.KubernetesSecrets) == 0 {
+		return true
+	}
+
+	for _, podSecret := range podSecrets {
+		for _, cachedSecret := range cachedCreds.KubernetesSecrets {
+
+			if podSecret.CredentialHash == cachedSecret.CredentialHash { // TODO: if we're salting, we need to rehash the podSecret with the cachedSecret salt
+				// TODO: should we record the new secret in case it has different coordinates? If the secret rotates, we will pull unnecessarily otherwise
+				return false
+			}
+
+			if podSecret.UID == cachedSecret.UID &&
+				podSecret.Namespace == cachedSecret.Namespace &&
+				podSecret.Name == cachedSecret.Name {
+				// TODO: should we record the new creds in this case so that we don't pull if these are present in a different secret?
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func (f *FileBasedImagePullManager) PruneUnknownRecords(imageList []string, until time.Time) {
+	f.pulledAccessors.GlobalLock()
+	defer f.pulledAccessors.GlobalUnlock()
+
+	var imageRecords []string
+	err := processDirFiles(f.pulledDir,
+		func(filePath string, fileContent []byte) error {
+			pullRecord := imagemanagerv1alpha1.ImagePulledRecord{}
+			if _, _, err := f.decoder.Decode(fileContent, nil, &pullRecord); err != nil {
+				klog.V(5).InfoS("failed to deserialize, skipping file", "filePath", filePath, "error", err)
+				return nil
+			}
+
+			if pullRecord.LastUpdatedTime.Time.Before(until) {
+				imageRecords = append(imageRecords, pullRecord.ImageRef)
+			}
+			return nil
+		})
+
+	if err != nil {
+		klog.ErrorS(err, "failed to garbage collect ImagePulledRecord files")
+		return
+	}
+
+	imagesInUse := sets.New(imageList...)
+	for _, ir := range imageRecords {
+		if imagesInUse.Has(ir) {
+			continue
+		}
+
+		recordToRemovePath := filepath.Join(f.pulledDir, cacheFilename(ir))
+		if err := os.Remove(recordToRemovePath); err != nil {
+			klog.ErrorS(err, "failed to remove an ImagePulledRecord file", "filePath", recordToRemovePath)
+		}
+	}
+
+}
+
+// startupHousekeeping gathers all the images from pull intent records that exist
+// from the previous kubelet runs.
+// If the CRI reports any of the above images as already pulled, we turn the
+// pull intent into a pulled record and the original pull intent is deleted.
+//
+// This method is not thread-safe and it should only be called upon the creation
+// of the FileBasedImagePullManager.
+func (f *FileBasedImagePullManager) startupHousekeeping(ctx context.Context) {
+	inFlightPulls := make(map[string]string) // image -> path
+
+	// walk the pulling directory for any pull intent records
+	err := processDirFiles(f.pullingDir,
+		func(filePath string, fileContent []byte) error {
+			var intent imagemanagerv1alpha1.ImagePullIntent
+			if _, _, err := f.decoder.Decode(fileContent, nil, &intent); err != nil {
+				klog.ErrorS(err, "skipping file, failed to deserialize to ImagePullIntent", "filePath", filePath)
```

Comment:
should we skip or delete?

Reply:
my read is the only way this happens is kubelet reset while recording intent which should be before the pull request so.. delete?

Reply:
delete seems safe here, I'll switch to it